### PR TITLE
*: support roll update

### LINKED ISSUE #121: [feature] support rolling update
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

Upgrade the follower nodes first, and then upgrade the leader node last. 

For a two-node cluster, we need to adjust the parameter rpl_semi_sync_master_timeout before upgrading from the node. Otherwise, the leader node will hang due to semi-sync.

**Describe alternatives you've considered**

**Additional context**

Reference: [Qingcloud MySQL Plus](https://docs.qingcloud.com/product/database_cache/app_mysql_plus/index.html)

### PR DESCRIPTION
**Pull request by zhyass**
### What type of PR is this?

/enhancement

### Which issue(s) this PR fixes?

Fixes #121 

### What this PR does?

Summary:

Upgrade the follower nodes first, and then upgrade the leader node last.

### CODE REVIEW

**Review by andyli029 — COMMENTED** [review]
Need add logic to cluster about two nodes

**Review by andyli029 — COMMENTED** [review]
Ping @acekingke

File: sidecar/init.go [hunk 1]
```diff
@@ -75,9 +73,24 @@ func runInitCommand(cfg *Config) error {
...
+	clientConfig, err := cfg.buildClientConfig()
+	if err != nil {
+		return fmt.Errorf("failed to build client.conf: %s", err)
+	}
+	// save client.conf to /etc/mysql.
+	if err := clientConfig.SaveTo(path.Join(clientConfPath)); err != nil {
+		return fmt.Errorf("failed to save client.conf: %s", err)
+	}
+
+	if err = os.Mkdir(extraConfPath, os.FileMode(0755)); err != nil {
```
Comment (acekingke) [inline sidecar/init.go:91]: fileMode 0755

File: internal/sql_runner.go [hunk 2]
```diff
@@ -168,9 +168,42 @@ func (s *SQLRunner) RunQuery(query string, args ...interface{}) error {
...
-func (sr *SQLRunner) GetGlobalVariable(param string, val interface{}) error {
+func (s *SQLRunner) GetGlobalVariable(param string, val interface{}) error {
 	query := fmt.Sprintf("select @@global.%s", param)
-	return sr.db.QueryRow(query).Scan(val)
+	return s.db.QueryRow(query).Scan(val)
+}
+
+func (s *SQLRunner) CheckProcesslist() (bool, error) {
+	var rows *sql.Rows
+	rows, err := s.db.Query("show processlist;")
```
[the remark this answers is not in the document]
Comment (zhyass) [reply internal/sql_runner.go:178]: https://github.com/zhyass/radondb-mysql-kubernetes/blob/upgrade/cluster/syncer/statefulset.go#L271

**Comment by andyli029** [comment]
Fix :make run failed #149

ConfMapVolumeMountPath

File: utils/constants.go [hunk 3]
```diff
@@ -74,15 +74,18 @@ const (
 	InitFileVolumeName = "init-mysql"
 
 	// volumes mount path.
-	MyCnfMountPath          = "/etc/mysql/my.cnf"
```
Comment (andyli029) [inline utils/constants.go:77]: Fix #149

Comment (zhyass) [reply utils/constants.go:77]: https://github.com/radondb/radondb-mysql-kubernetes/pull/143/commits

File: cluster/syncer/statefulset.go [hunk 4]
```diff
@@ -17,98 +17,483 @@ limitations under the License.
...
+		if err != nil {
+			log.Error(err, "failed to check role", "pod", follower)
+			return false, nil
+		}
+		if status == corev1.ConditionTrue {
+			return true, nil
+		}
+
+		// If not leader, try to leader.
+		xenonHttpRequest(followerHost, "POST", "/v1/raft/trytoleader", rootPasswd, nil)
```
[the remark this answers is not in the document]
Comment (zhyass) [reply cluster/syncer/statefulset.go:306]: > When the XenonHttpRequest return error, should need retry ?

yes

File: testbin/setup-envtest.sh [hunk 5]
```diff
@@ -0,0 +1,96 @@
+#!/usr/bin/env bash
```
Comment (acekingke) [inline testbin/setup-envtest.sh:1]: testbin is generate by make test，should not need

**Comment by acekingke** [comment]
LGTM

**Pull request merged**